tests/test_qubit_device.py

```diff
@@ -824,3 +826,366 @@ def test_result_empty_tape(self, mock_qubit_device_with_paulis_and_methods, tol)
+        correct."""
+        params = np.array([theta, theta ** 3, np.sqrt(2) * theta])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.QubitStateVector(np.array([1.0, -1.0]) / np.sqrt(2), wires=0)
+            qml.Rot(*params, wires=[0])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        calculated_val = dev.adjoint_jacobian(tape)
+
+        # compare to finite differences
+        numeric_val = tape.jacobian(dev, method="numeric")
+        assert np.allclose(calculated_val, numeric_val, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("par", [1, -2, 1.623, -0.051, 0])  # integers, floats, zero
+    def test_ry_gradient(self, par, tol, dev):
+        """Test that the gradient of the RY gate matches the exact analytic formula."""
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RY(par, wires=[0])
+            qml.expval(qml.PauliX(0))
+
+        tape.trainable_params = {0}
+
+        # gradients
+        exact = np.cos(par)
+        grad_F = tape.jacobian(dev, method="numeric")
+        grad_A = dev.adjoint_jacobian(tape)
+
+        # different methods must agree
+        assert np.allclose(grad_F, exact, atol=tol, rtol=0)
+        assert np.allclose(grad_A, exact, atol=tol, rtol=0)
+
+    def test_rx_gradient(self, tol, dev):
+        """Test that the gradient of the RX gate matches the known formula."""
+        a = 0.7418
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(a, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        # circuit jacobians
+        dev_jacobian = dev.adjoint_jacobian(tape)
+        expected_jacobian = -np.sin(a)
+        assert np.allclose(dev_jacobian, expected_jacobian, atol=tol, rtol=0)
+
+    def test_multiple_rx_gradient(self, tol):
+        """Tests that the gradient of multiple RX gates in a circuit yields the correct result."""
+        dev = qml.device("default.qubit", wires=3)
+        params = np.array([np.pi, np.pi / 2, np.pi / 3])
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(params[0], wires=0)
+            qml.RX(params[1], wires=1)
+            qml.RX(params[2], wires=2)
+
+            for idx in range(3):
+                qml.expval(qml.PauliZ(idx))
+
+        # circuit jacobians
+        dev_jacobian = dev.adjoint_jacobian(tape)
+        expected_jacobian = -np.diag(np.sin(params))
+        assert np.allclose(dev_jacobian, expected_jacobian, atol=tol, rtol=0)
+
+    qubit_ops = [getattr(qml, name) for name in qml.ops._qubit__ops__]
+    analytic_qubit_ops = {cls for cls in qubit_ops if cls.grad_method == "A"}
+    analytic_qubit_ops -= {
+        qml.CRot,  # not supported for RewindTape
+        qml.PauliRot,  # not supported in test
+        qml.MultiRZ,  # not supported in test
+        qml.U1,  # not supported on device
+        qml.U2,  # not supported on device
+        qml.U3,  # not supported on device
```

Comment:
These should work if using a QNode, since they will be decomposed for default.qubit

Reply:
Right! I had put those comments when using a tape (which didn't decompose).
These 3 are also problematic since they are decomposed into gates that combine the parameters.